pravega-sensor-collector/src/main/java/io/pravega/sensor/collector/opcua/OpcUaClientDriver.java

```diff
@@ -0,0 +1,211 @@
+            if (opcUaClient.getAddressSpace().getNode(getNodeID()).getNodeClass().getValue() == NodeClass.Variable.getValue())
+            {
+                sensorList.add(getNodeID());
+            }
+            filterNode(opcUaClient,getNodeID());
+            readValueIds = sensorList.stream().map(nodeId -> new ReadValueId(nodeId, AttributeId.Value.uid(), null, null)).collect(Collectors.toUnmodifiableList());
+        } catch (UaException | InterruptedException | ExecutionException e) {
+            log.error("Error in connection to UA Server", e);
+        }
+    }
+
+
+    @Override
+    public List<OpcUaRawData> readRawData() throws ExecutionException, InterruptedException {
+        List<OpcUaRawData> dataList = new LinkedList<>();
+        ReadResponse readResp = opcUaClient.read(0, TimestampsToReturn.Source,readValueIds).get();
+        DataValue[] aggregatedData = readResp.getResults();
+        int i=0;
+        for (DataValue data : aggregatedData)
+        {
+            // As bulk read operation is in-place read , the list ordering of the input nodes will match the data fetched from responses.
+            Variant rawVariant = data.getValue();
+            String dataTypeClass = TypeUtil.getBackingClass(rawVariant.getDataType().get().toNodeId(ns).get()).getName();
+            log.trace("Sensor name {} : Raw Data {}",sensorList.get(i).getIdentifier(),rawVariant.getValue());
+            dataList.add(new OpcUaRawData(rawVariant.getValue(),data.getSourceTime().getUtcTime(), sensorList.get(i++).getIdentifier().toString(),dataTypeClass));
+        }
+        return dataList;
+    }
+
+    private void filterNode(OpcUaClient client, NodeId rootNode) {
+        BrowseDescription browse = new BrowseDescription(
+                rootNode,
+                BrowseDirection.Forward,
+                Identifiers.References,
+                true,
+                uint(NodeClass.Object.getValue() | NodeClass.Variable.getValue()), // Get both Objects and Variable types while browsing
+                uint(BrowseResultMask.All.getValue())
+        );
+
+        try {
+            BrowseResult browseResult = client.browse(browse).get();
+            List<ReferenceDescription> references = toList(browseResult.getReferences());
+
+            for (ReferenceDescription rd : references) {
+                if (rd.getBrowseName().getName().equalsIgnoreCase("_Hints"))
+                {
+                    //Skip node iteration if the node name is _Hints as it contains hints about variables creation functions supported by server.
+                    continue;
+                }
+                else if (nodeFilter.matcher(rd.getBrowseName().getName()).find() && rd.getNodeClass().getValue() == NodeClass.Variable.getValue())
+                {
+                    //Sensor which matches RegEx and node type being a Variable.
+                    log.info("Qualified Sensor:"+rd.getNodeId().toNodeId(ns).get().getIdentifier());
```

Comment:
Avoid constructing strings when logging. This causes unnecessary computation when the logging doesn't occur. Use `{}` placeholders.

Reply:
Amended.